File: mm-rs/src/jsonc/scanner.rs
#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    EOF,
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    Colon,
    Comma,
    String,
    Number,
    True,
    False,
    Null,
    Comment,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
    pub line: usize,
    pub column: usize,
}

pub struct Scanner {
    input: Vec<char>,
    position: usize,
    line: usize,
    column: usize,
}

impl Scanner {
    pub fn new(input: &str) -> Self {
        Self {
            input: input.chars().collect(),
            position: 0,
            line: 1,
            column: 1,
        }
    }

    pub fn next_token(&mut self) -> Token {
        self.skip_whitespace();

        if self.position >= self.input.len() {
            let token = Token {
                token_type: TokenType::EOF,
                literal: String::new(),
                line: self.line,
                column: self.column,
            };
            return token;
        }

        let ch = self.input[self.position];

        let token = if ch == '/' {
            self.scan_comment()
        } else {
            let t = match ch {
                '{' => self.create_token(TokenType::LBrace),
                '}' => self.create_token(TokenType::RBrace),
                '[' => self.create_token(TokenType::LBracket),
                ']' => self.create_token(TokenType::RBracket),
                ':' => self.create_token(TokenType::Colon),
                ',' => self.create_token(TokenType::Comma),
                '"' => self.scan_string(),
                _ if ch.is_ascii_digit() || ch == '-' => self.scan_number(),
                _ if ch.is_alphabetic() => self.scan_identifier(),
                _ => panic!(
                    "unexpected character: {} at line {}, column {}",
                    ch, self.line, self.column
                ),
            };
            t
        };

        token
    }

    fn create_token(&mut self, token_type: TokenType) -> Token {
        let token = Token {
            token_type,
            literal: String::new(),
            line: self.line,
            column: self.column,
        };
        self.advance(1);
        token
    }

    fn scan_comment(&mut self) -> Token {
        if self.position + 1 >= self.input.len() || self.input[self.position + 1] != '/' {
            return self.create_token(TokenType::Comment);
        }

        let start_line = self.line;
        let start_column = self.column;

        self.advance(2);
        let content_start = self.position;
        while self.position < self.input.len() && self.input[self.position] != '\n' {
            self.advance(1);
        }
        let literal: String = self.input[content_start..self.position].iter().collect();
        let literal = literal.trim().to_string();

        Token {
            token_type: TokenType::Comment,
            literal,
            line: start_line,
            column: start_column,
        }
    }

    fn scan_string(&mut self) -> Token {
        let start_line = self.line;
        let start_column = self.column;
        self.advance(1);

        let mut sb = String::new();
        while self.position < self.input.len() && self.input[self.position] != '"' {
            if self.input[self.position] == '\\' && self.position + 1 < self.input.len() {
                self.advance(1);
                let escaped = self.input[self.position];
                match escaped {
                    'n' => sb.push('\n'),
                    'r' => sb.push('\r'),
                    't' => sb.push('\t'),
                    'b' => sb.push('\x08'),
                    'f' => sb.push('\x0c'),
                    '"' => sb.push('"'),
                    '\\' => sb.push('\\'),
                    'u' => {
                        if self.position + 4 < self.input.len() {
                            self.advance(1);
                            let hex: String = self.input[self.position..self.position + 4]
                                .iter()
                                .collect();
                            if let Ok(unicode) = u32::from_str_radix(&hex, 16) {
                                sb.push(char::from_u32(unicode).unwrap_or('\u{FFFD}'));
                                self.advance(3);
                            }
                        }
                    }
                    _ => sb.push(escaped),
                }
            } else {
                if self.input[self.position] == '\n' {
                    self.line += 1;
                    self.column = 0;
                }
                sb.push(self.input[self.position]);
            }
            self.advance(1);
        }

        if self.position < self.input.len() {
            self.advance(1);
        }

        Token {
            token_type: TokenType::String,
            literal: sb,
            line: start_line,
            column: start_column,
        }
    }

    fn scan_number(&mut self) -> Token {
        let start_line = self.line;
        let start_column = self.column;
        let mut sb = String::new();

        if self.input[self.position] == '-' {
            sb.push('-');
            self.advance(1);
        }

        while self.position < self.input.len() {
            let ch = self.input[self.position];
            if ch.is_ascii_digit() || ch == '.' || ch == 'e' || ch == 'E' || ch == '+' || ch == '_'
            {
                if ch == '_' {
                    self.advance(1);
                    continue;
                }
                sb.push(ch);
                self.advance(1);
            } else {
                break;
            }
        }

        if self.position < self.input.len() {
            let ch = self.input[self.position];
            if ch == 'f' || ch == 'F' {
                sb.push(ch);
                self.advance(1);
            }
        }

        Token {
            token_type: TokenType::Number,
            literal: sb,
            line: start_line,
            column: start_column,
        }
    }

    fn scan_identifier(&mut self) -> Token {
        let start_line = self.line;
        let start_column = self.column;
        let mut sb = String::new();

        while self.position < self.input.len()
            && (self.input[self.position].is_alphanumeric() || self.input[self.position] == '_')
        {
            sb.push(self.input[self.position]);
            self.advance(1);
        }

        let identifier = sb.to_lowercase();
        let (token_type, literal) = match identifier.as_str() {
            "true" => (TokenType::True, "true".to_string()),
            "false" => (TokenType::False, "false".to_string()),
            "null" => (TokenType::Null, "null".to_string()),
            _ => (TokenType::String, sb.clone()),
        };

        Token {
            token_type,
            literal,
            line: start_line,
            column: start_column,
        }
    }

    fn skip_whitespace(&mut self) {
        while self.position < self.input.len() {
            let ch = self.input[self.position];
            if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
                self.advance(1);
            } else {
                break;
            }
        }
    }

    fn advance(&mut self, count: usize) {
        for _ in 0..count {
            if self.position < self.input.len() {
                if self.input[self.position] == '\n' {
                    self.line += 1;
                    self.column = 0;
                } else {
                    self.column += 1;
                }
                self.position += 1;
            }
        }
    }
}

File: mm-cpp/src/jsonc/scanner.cpp
#include "scanner.hpp"
#include <cctype>
#include <stdexcept>

namespace mmc {
namespace jsonc {

char Scanner::peek() const {
  if (position_ >= input_.size())
    return '\0';
  return input_[position_];
}

void Scanner::advance(size_t count) {
  for (size_t i = 0; i < count; ++i) {
    if (position_ >= input_.size())
      break;
    if (input_[position_] == '\n') {
      ++line_;
      column_ = 0;
    } else {
      ++column_;
    }
    ++position_;
  }
}

void Scanner::skipWhitespace() {
  while (position_ < input_.size()) {
    char ch = input_[position_];
    if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r') {
      advance();
    } else {
      break;
    }
  }
}

Token Scanner::scanString() {
  size_t startLine = line_;
  size_t startCol = column_;
  advance();
  std::string sb;
  while (position_ < input_.size() && input_[position_] != '"') {
    if (input_[position_] == '\\' && position_ + 1 < input_.size()) {
      sb.push_back(input_[position_]);
      advance();
      sb.push_back(input_[position_]);
    } else {
      sb.push_back(input_[position_]);
    }
    advance();
  }
  if (position_ < input_.size() && input_[position_] == '"')
    advance();
  return Token(TokenType::String, sb, startLine, startCol);
}

Token Scanner::scanComment() {
  size_t startLine = line_;
  size_t startCol = column_;
  advance();
  if (position_ >= input_.size())
    return Token(TokenType::Comment, "", startLine, startCol);

  if (input_[position_] == '/') {
    advance();
    std::string comment;
    while (position_ < input_.size() && input_[position_] != '\n') {
      comment.push_back(input_[position_]);
      advance();
    }
    // Trim whitespace (matching Go's strings.TrimSpace)
    size_t start = 0;
    while (start < comment.size() &&
           (comment[start] == ' ' || comment[start] == '\t' ||
            comment[start] == '\r' || comment[start] == '\n'))
      ++start;
    size_t end = comment.size();
    while (end > start &&
           (comment[end - 1] == ' ' || comment[end - 1] == '\t' ||
            comment[end - 1] == '\r' || comment[end - 1] == '\n'))
      --end;
    comment = comment.substr(start, end - start);
    return Token(TokenType::Comment, comment, startLine, startCol);
  }
  return Token(TokenType::Comment, "", startLine, startCol);
}

Token Scanner::scanNumber() {
  size_t startLine = line_;
  size_t startCol = column_;
  std::string sb;
  if (input_[position_] == '-') {
    sb.push_back('-');
    advance();
  }
  while (position_ < input_.size()) {
    char ch = input_[position_];
    if (std::isdigit(static_cast<unsigned char>(ch)) || ch == '.' ||
        ch == 'e' || ch == 'E' || ch == '+' || ch == '_') {
      if (ch == '_') {
        advance();
        continue;
      }
      sb.push_back(ch);
      advance();
    } else {
      break;
    }
  }
  return Token(TokenType::Number, sb, startLine, startCol);
}

Token Scanner::scanIdentifier() {
  size_t startLine = line_;
  size_t startCol = column_;
  std::string sb;
  while (position_ < input_.size() &&
         (std::isalnum(static_cast<unsigned char>(input_[position_])) ||
          input_[position_] == '_')) {
    sb.push_back(input_[position_]);
    advance();
  }
  if (sb == "true")
    return Token(TokenType::True, sb, startLine, startCol);
  if (sb == "false")
    return Token(TokenType::False, sb, startLine, startCol);
  if (sb == "null")
    return Token(TokenType::Null, sb, startLine, startCol);
  return Token(TokenType::String, sb, startLine, startCol);
}

Token Scanner::nextToken() {
  skipWhitespace();
  if (position_ >= input_.size())
    return createToken(TokenType::EOF_, "");

  char ch = input_[position_];
  size_t startLine = line_;
  size_t startCol = column_;

  if (ch == '/')
    return scanComment();

  switch (ch) {
  case '{':
    advance();
    return Token(TokenType::LBrace, "", startLine, startCol);
  case '}':
    advance();
    return Token(TokenType::RBrace, "", startLine, startCol);
  case '[':
    advance();
    return Token(TokenType::LBracket, "", startLine, startCol);
  case ']':
    advance();
    return Token(TokenType::RBracket, "", startLine, startCol);
  case ':':
    advance();
    return Token(TokenType::Colon, "", startLine, startCol);
  case ',':
    advance();
    return Token(TokenType::Comma, "", startLine, startCol);
  case '"':
    return scanString();
  default:
    if (std::isdigit(static_cast<unsigned char>(ch)) || ch == '-')
      return scanNumber();
    else if (std::isalpha(static_cast<unsigned char>(ch)))
      return scanIdentifier();
    advance();
    return Token(TokenType::Comment, "", startLine, startCol);
  }
}

} // namespace jsonc
} // namespace mmc